src/hivpy/sexual_behaviour.py

```diff
@@ -109,37 +110,48 @@ def prob_transition(self, sex, rred, i, j):
 
         return Probability
 
+    def get_partners_for_group(self, args):
+        """Calculates the number of short term partners for people in a given intersection of
+        sex and sexual behaviour group. Args are: [sex, sexual behaviour group, size of group]"""
+        sex = args[0]
+        group = int(args[1])
+        size = args[2]
+        return self.short_term_partners[sex][group].sample(size)
+
     def num_short_term_partners(self, population: pd.DataFrame):
-        for sex in SexType:
-            for g in SexBehaviours[sex]:
-                index = selector(population, sex=(operator.eq, sex),
-                                 sex_behaviour=(operator.eq, g),
-                                 age=(operator.gt, 15))
-                population.loc[index, col.NUM_PARTNERS] = (
-                    self.short_term_partners[sex][g].sample(size=sum(index)))
+        """Calculate the number of short term partners for the whole population"""
+        # can we avoid doing population.loc[active_pop] twice? Does it waste time?
+        active_pop = population.index[(15 <= population.age) & (population.age < 65)]
```

Comment:
This may be a useful method on the `Population`, since it's also computed when updating sex groups. Maybe a method that also does the `.loc` and returns the reduced dataframe directly.